client/ayon_houdini/plugins/publish/extract_last_published.py

```diff
@@ -0,0 +1,89 @@
+from ayon_houdini.api import plugin
+
+
+class ExtractLastPublished(plugin.HoudiniExtractorPlugin):
+    """
+    Generic Extractor that copies files from last published
+    to staging directory.
+    It works only if instance data includes "last_version_published_files"
+    and there are frames to fix.
+
+    The files from last published are base of files which will be extended/fixed for specific
+    frames.
+    """
+
+    order = pyblish.api.ExtractorOrder - 0.1
+    label = "Extract Last Published"
+    targets = ["local"]  # Same target as `CollectFramesFixDef`
+    families = ["*"]
+
+    def process(self, instance):
+        frames_to_fix = instance.data.get("frames_to_fix")
+        last_published = instance.data.get("last_version_published_files")
+        if not last_published:
+            self.log.debug("Skipping, No last publish found.")
+            return 
+        if not frames_to_fix :
+            self.log.debug("Skipping, No frames to fix.")
+            return
+        last_published_and_frames = collect_frames(last_published)
+        
+        if not all(last_published_and_frames.values()):
+            self.log.debug("Skipping, No file sequence found in the "
+                           "last version published files.")
+            return
+
+        expected_filenames = []
+        staging_dir = instance.data.get("stagingDir")
+        expectedFiles = instance.data.get("expectedFiles", [])
+
+        # 'expectedFiles' are preferred over 'frames'
+        if expectedFiles:
+            # Products with expected files
+            # This can be Render products or submitted cache to farm.
+            for expected in expectedFiles:
+                # expected.values() is a list of lists
+                expected_filenames.extend(sum(expected.values(), []))
+        else:
+            # Products with frames or single file.
+            frames = instance.data.get("frames", "")
+            if isinstance(frames, str):
+                # single file.
+                expected_filenames.append("{}/{}".format(staging_dir, frames))
+            else:
+                # list of frame.
+                expected_filenames.extend(
+                    ["{}/{}".format(staging_dir, f) for f in frames]
+                )
+
+        os.makedirs(staging_dir, exist_ok=True)
+
+        expected_and_frames = collect_frames(expected_filenames)
+        frames_and_expected = {v: k for k, v in expected_and_frames.items()}
+        frames_to_fix = clique.parse(frames_to_fix, "{ranges}")
+        
+        anatomy = instance.context.data["anatomy"]
+
+        for file_path, frame in last_published_and_frames.items():
+            if frame is None:
+                continue
+            file_path = anatomy.fill_root(file_path)
+            if not os.path.exists(file_path):
+                continue
+            target_file_name = frames_and_expected.get(frame)
+            if not target_file_name:
+                continue
```

Comment:
Same here... if we can't define where it ends up we ignore it silently? Does that seem right? It's collected into `last_version_published_files` data because it was data we needed to transfer, right? Or did it just contain all last version's files and we're filtering here.

I have a feeling we need a generic way (that hopefully also is not houdini-specific!) to:

- Define the frames we want to re-render/re-export in the current publish -> The frames to render
- Those not include but still within frame start and frame end of the instance are transferred from last version -> The list we need to copy from last version.
- In validation we need to ensure the frames to copy from last version actually exist (so it actually makes sense to try and ignore those and we're safe to continue?)

Reply:
I agree, I tried to make it generic although I was mostly stealing code from Nuke.

Reply:
What is the case when those files don't exist ?
- wrong `frames to fix` input ? 
- Or  the published files don't exist ? i.e. they were deleted manually without removing them from the database ? 
- Or a storage connection error resulting in making the files look like they are missed ? which I assume will result in an error anyway during the publishing.

Reply:
> I agree, I tried to make it generic although I was mostly stealing code from Nuke.

When doing such a copy - it's good to mark that very clear in the code... especially with such a feature that should really be taken OUT of Nuke and into core... Because this plug-in should be generic, not to Houdini, not to Nuke, but something that can all share.